packages/tupaia-web-server/src/routes/MapOverlaysRoute.ts

```diff
@@ -0,0 +1,172 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2023 Beyond Essential Systems Pty Ltd
+ */
+
+import { Request } from 'express';
+import { Route } from '@tupaia/server-boilerplate';
+import { MapOverlay, MapOverlayGroup, MapOverlayGroupRelation } from '@tupaia/types';
+import groupBy from 'lodash.groupby';
+import keyBy from 'lodash.keyby';
+
+// TODO: WAITP-1278 split request types to types package
+// (And actually define it)
+export type MapOverlaysRequest = Request<any, any, any, any>;
+
+// TODO: Can these be moved into types?
+const ROOT_MAP_OVERLAY_CODE = 'Root';
+const MAP_OVERLAY_CHILD_TYPE = 'mapOverlay';
+
+// We return a simplified version of data to the frontend
+interface TranslatedMapOverlay {
+  code: string;
+  name: string;
+  reportCode: string;
+  legacy: boolean;
+  // ...config
+}
+interface TranslatedMapOverlayGroup {
+  name: string;
+  children: OverlayChild[];
+}
+type OverlayChild = TranslatedMapOverlayGroup | TranslatedMapOverlay;
+
+export class MapOverlaysRoute extends Route<MapOverlaysRequest> {
+  public async buildResponse() {
+    const { params, ctx } = this.req;
+    const { projectCode, entityCode } = params;
```

Comment:
You make a great point about how we do the map overlay nesting backwards. It definitely makes this kind of route pretty confusing. And it's inconsistent with dashboard relations.

We took a stab at refactoring this [here](https://linear.app/bes/issue/RN-839/manage-permissions-and-projects-for-map-overlays-via-map-overlay) but at the time refactor didn't seem worth the effort (and honestly I just lost heart and got a bit distracted with more pressing matters).

It would be nice to finish the job really. Maybe once the epic branch gets merged we can look at picking it up in a maintenance sprint.